### PR TITLE
Pass xml date types as javascript dates to custom functions

### LINKED ISSUE #155: Custom XPath functions that take in an xs:date or xs:dateTime do not get given a Javascript Date
Example function:
```
registerCustomXPathFunction(
            'fn:days-from-date',
            ['xs:date'],
            'xs:integer',
            (context: any, date: Date) => {
                return Math.floor(date.getTime() / 8.64e7)
            })
```

> TypeError: date.getTime is not a function

It returns this instead:
`w {
        h: 1970,
        a: 1,
        j: 1,
        o: 0,
        V: 0,
        C: 0,
        sa: 0,
        O: null,
        type: 'xs:date' }`

### PR DESCRIPTION
**Pull request by ymor**
Fixes #155
I hope you don't mind, but i also brought in a union type for the xml/xpath/xquery data types. I've not updated all places where it could be used and are probably missing a few types. I only covered just enough to get typescript to build.

### CODE REVIEW

**Comment by DrRataplan** [comment]
Great work, thanks! The union type makes sense too, for the xml schema related types!

I'll review the PR tomorrow, @JosVerburg could you also take a look?

File: src/expressions/dataTypes/Value.ts [hunk 1]
```diff
@@ -1,3 +1,42 @@
 export default class Value {
-	constructor(public type: string, readonly value: any) {}
+	constructor(public type: ValueType, readonly value: any) {}
 }
+
+export type ValueType =
```
Comment (JosVerburg) [inline src/expressions/dataTypes/Value.ts]: Very nice!
I think we can even split these up in atomic specific and all types, or does that cause more issues? e.g.
```js
export type AnyAtomicType =
	| 'xs:boolean'
	| 'xs:string' ...;
export type ValueType =
	AnyAtomicType
	| 'node()'
	| 'element()'
	| 'document()'
	| 'text'
	| 'comment()'
	| 'item()'
	| 'function(*)'
	| 'map(*)'
	| 'array(*)';
```

Comment (ymor) [reply src/expressions/dataTypes/Value.ts]: You certainly can, but I'm not sure on the rules :) 
Would the idea be that the AtomicValue type could only take an AnyAtomicType? And Value would still take ValueType?

Comment (JosVerburg) [reply src/expressions/dataTypes/Value.ts]: Yes, that is indeed my idea. That way we prevent non atomic values ending up in an `AtomicValue`.
It doesn't have to be part of this PR btw, we can of course do further improvements like those sometime later.

Comment (ymor) [reply src/expressions/dataTypes/Value.ts]: Cool, well i had a quick go and it does cascade/get out of hand pretty quickly! So probably one for later :)

Comment (DrRataplan) [reply src/expressions/dataTypes/Value.ts]: Cool, something for later it is!

**Comment by DrRataplan** [comment]
Thanks for your contribution! We will release this later this week, when the other big PR has come through

**Pull request merged**